xapian-applications/omega/handler_libextractor.cc

```diff
@@ -0,0 +1,120 @@
+/** @file handler_libextractor.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <extractor.h>
+
+#include <cstring>
+
+using namespace std;
+
+string title, keywords, author, pages;
+
+/*
+* cls: passed as last parameter from EXTRACTOR_extract
+* plugin_name:	name of the plugin
+* type: mime-type of file according to libextractor
+* format: format information about data
+* data_mime_type: mimetype of data according to libextractor
+* data: actual meta-data found
+* data_len: number of bytes in data
+*/
+static int
+process_metadata(void* cls,
+		 const char* plugin_name,
+		 enum EXTRACTOR_MetaType type,
+		 enum EXTRACTOR_MetaFormat format,
+		 const char* data_mime_type,
+		 const char* data,
+		 size_t data_len)
+{
+    switch (format) {
+	case EXTRACTOR_METAFORMAT_UTF8:
+	    break;
+	default:
+	    // specific encoding unknown
+	    // EXTRACTOR_METAFORMAT_UNKNOWN
+	    // EXTRACTOR_METAFORMAT_BINARY
+	    // EXTRACTOR_METAFORMAT_C_STRING
+	    return 0;
+    }
+
+    switch (type) {
+	case EXTRACTOR_METATYPE_TITLE:
+	case EXTRACTOR_METATYPE_BOOK_TITLE:
+	case EXTRACTOR_METATYPE_SUBJECT:
+	    title = data;
```

Comment:
I couldn't find if `data` is null terminated on libextractor docs. Maybe using `data_len` would be a good idea.
There is actually a [string constructor](http://www.cplusplus.com/reference/string/string/string/) that uses it(see constructor number 5).

Reply:
Thanks for pointing this out. I have changed to code and now it uses the string constructor 

`string (const char* s, size_t n);`